### PR TITLE
feat(in-mail-service): Implementation of In-mail service

### PR DESCRIPTION
**Pull request by bhavyasf**
SFO-141

## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Intermediate change (work in progress)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist:

- [x] Performed a self-review of my own code
- [x] npm test passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the style guide
- [ ] API Documentation in code was updated
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by samarpan-b** [comment]
@bhavyasf please rebas eit. Also test cases are failing. please check

File: services/in-mail-service/README.md [hunk 1]
```diff
@@ -1,3 +1,691 @@
...
+   npm i @sourceloop/inmail-service
+```
+
+## Implementation
+
+### Migration
+First of all we need to set up migration for the in-mail database.
+The query for Migration is given below.
+
+```sql
```
Comment (samarpan-b) [inline services/in-mail-service/README.md:20]: why this ? Put migration in a separate folder as executable scripts like we do elsewhere

File: services/in-mail-service/migrations/database.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
+{
```
Comment (samarpan-b) [inline services/in-mail-service/migrations/database.json:1]: use latest lb4 cli. in that this file wont be needed

File: services/in-mail-service/src/controllers/collector.controller.ts [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+    @repository(GroupRepository) public groupRepository: GroupRepository,
+    @repository(ThreadRepository) public threadRepository: ThreadRepository,
+    @repository(AttachmentRepository)
+    public attachmentRepository: AttachmentRepository,
+    @inject(AuthenticationBindings.CURRENT_USER)
+    public user: IAuthUserWithPermissions,
+  ) {}
+
+  @authenticate(STRATEGY.BEARER)
+  @authorize(['*'])
```
Comment (samarpan-b) [inline services/in-mail-service/src/controllers/collector.controller.ts:54]: why * ?

File: services/in-mail-service/src/controllers/collector.controller.ts [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+    responses: {
+      200: {
+        description:
+          'Fetches a thread along with message, group, attachment(s) etc based on unique thread Id',
+        content: {
+          'application/json': {
+            schema: {
+              type: 'object',
+              title: 'MailResponse',
+              properties: {
```
Comment (samarpan-b) [inline services/in-mail-service/src/controllers/collector.controller.ts:67]: why properties like this ? use LB4 decorators

File: services/in-mail-service/src/controllers/collector.controller.ts [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+              version: {type: 'string', example: '20200101'},
+            },
+          },
+        },
+      } as ResponseObject,
+      403: {description: FORBIDDEN_ERROR_MESSAGE},
+    },
+  })
+  async fetchThreadList(
+    @param.path.string('version') version: string,
```
Comment (samarpan-b) [inline services/in-mail-service/src/controllers/collector.controller.ts:247]: use Filter provided by LB4

File: services/in-mail-service/src/controllers/collector.controller.ts [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+        },
+      } as ResponseObject,
+      403: {description: FORBIDDEN_ERROR_MESSAGE},
+    },
+  })
+  async fetchInboxList(
+    @param.path.string('version') version: string,
+    @param.path.number('offset') offset: number,
+
+    @param.path.number('limit') limit: number,
```
Comment (samarpan-b) [inline services/in-mail-service/src/controllers/collector.controller.ts:364]: Please do LB4 way.

File: services/in-mail-service/src/controllers/collector.controller.ts [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+                }),
+              },
+            },
+          },
+        },
+      },
+    },
+  })
+  async fetchMetaData(
+    @param.path.string('messageId') messageId: string,
```
Comment (samarpan-b) [inline services/in-mail-service/src/controllers/collector.controller.ts:473]: All the GET APIs have incorrect signatures. Please make them similar to LB4 standards

File: services/in-mail-service/src/controllers/originator.controller.ts [hunk 8]
```diff
@@ -0,0 +1,839 @@
...
+            schema: {$ref: ID_RESPONSE_SCHEMA},
+          },
+        },
+      } as ResponseObject,
+      403: {description: FORBIDDEN_ERROR_MESSAGE},
+      404: {description: NOT_FOUND_MESSAGE},
+    },
+  })
+  async composeMail(
+    @requestBody({
```
Comment (samarpan-b) [inline services/in-mail-service/src/controllers/originator.controller.ts:89]: same comment as in other controller

File: services/in-mail-service/src/helpers/repository.helper.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
+import {Filter} from '@loopback/repository';
+import {Message, Group, Thread, Meta, Attachment} from '../models';
+
+type FilterTypes = Message | Group | Thread | Meta | Attachment;
+
+export const repositoryHelper = {
```
Comment (samarpan-b) [inline services/in-mail-service/src/helpers/repository.helper.ts:6]: no need of this. use services for business logic if needed

File: services/in-mail-service/src/models/group.model.ts [hunk 10]
```diff
@@ -0,0 +1,117 @@
+import {model, property, belongsTo, AnyObject} from '@loopback/repository';
+import {Message} from './message.model';
+import {Thread} from './thread.model';
+import {
+  UserModifiableEntity,
+  ExternalIdentifierEnabledEntity,
+} from '@sourceloop/core';
+
+export enum VisibilityMarker {
```
Comment (samarpan-b) [inline services/in-mail-service/src/models/group.model.ts:9]: keep enums in separate files

File: services/in-mail-service/src/repositories/attachment.repository.ts [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+  DataObject,
+  juggler,
+} from '@loopback/repository';
+import {Attachment, Message} from '../models';
+import {inject, Getter} from '@loopback/core';
+import {AuthenticationBindings} from 'loopback4-authentication';
+import {IAuthUserWithPermissions} from '@sourceloop/core';
+import {repositoryHelper} from '../helpers';
+
+export class AttachmentRepository extends DefaultCrudRepository<
```
Comment (samarpan-b) [inline services/in-mail-service/src/repositories/attachment.repository.ts:15]: extend from usermodifyrepo

File: services/in-mail-service/src/repositories/attachment.repository.ts [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    entity.createdBy = currentUser?.id;
+    entity.createdOn = new Date();
+    return super.create(entity, options);
+  }
+  async updateById(
+    id: string,
+    entity: DataObject<Attachment>,
+    options?: AnyObject,
+  ): Promise<void> {
+    const currentUser = await this.getCurrentUser();
```
Comment (samarpan-b) [inline services/in-mail-service/src/repositories/attachment.repository.ts:47]: not needed. use different base class

File: services/in-mail-service/src/repositories/message.repository.ts [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+    const user = await this.getCurrentUser();
+    entity.modifiedBy = user?.id;
+    entity.modifiedOn = new Date();
+    return super.update(entity, options);
+  }
+  async find(
+    filter: Filter<Message> | undefined,
+    options?: AnyObject,
+  ): Promise<(Message & MessageRelations)[]> {
+    repositoryHelper.addFalseDeletedConditionInInclude(filter);
```
Comment (samarpan-b) [inline services/in-mail-service/src/repositories/message.repository.ts:173]: I dont understand this at all. Why is it done like this ?

**Review by samarpan-b — CHANGES_REQUESTED** [review]

File: services/in-mail-service/src/helpers/repository.helper.ts (hunk 9 above)
Comment (bhavyasf) [reply services/in-mail-service/src/helpers/repository.helper.ts:6]: Ok. Basically the purpose of this helper was to fetch only data in default where `deleted = false`

File: services/in-mail-service/src/models/group.model.ts (hunk 10 above)
Comment (bhavyasf) [reply services/in-mail-service/src/models/group.model.ts:9]: ok!

File: services/in-mail-service/src/repositories/message.repository.ts (hunk 13 above)
Comment (bhavyasf) [reply services/in-mail-service/src/repositories/message.repository.ts:173]: Only used to fetch data in which `deleted = false` and for GET APIs  removing the deleted attribute

File: services/in-mail-service/src/repositories/attachment.repository.ts (hunk 11 above)
Comment (bhavyasf) [reply services/in-mail-service/src/repositories/attachment.repository.ts:15]: Ok!

File: services/in-mail-service/src/repositories/message.repository.ts (hunk 13 above)
Comment (samarpan-b) [reply services/in-mail-service/src/repositories/message.repository.ts:173]: There is a base repo for that already. No custom work needed.

File: services/in-mail-service/README.md (hunk 1 above)
Comment (bhavyasf) [reply services/in-mail-service/README.md:20]: Already added. I think I just need to be removed

File: services/in-mail-service/src/controllers/collector.controller.ts (hunk 5 above)
Comment (bhavyasf) [reply services/in-mail-service/src/controllers/collector.controller.ts:247]: version comes from the API request parameter that's why I used params.path

File: services/in-mail-service/migrations/database.json (hunk 2 above)
Comment (bhavyasf) [reply services/in-mail-service/migrations/database.json:1]: that's used for migrations using `db-migrate`

File: services/in-mail-service/src/repositories/attachment.repository.ts (hunk 12 above)
Comment (bhavyasf) [reply services/in-mail-service/src/repositories/attachment.repository.ts:47]: can you tell me which one to use?

Comment (samarpan-b) [reply services/in-mail-service/src/repositories/attachment.repository.ts:47]: check repositories in auth micro service

File: services/in-mail-service/src/index.ts [hunk 14]
```diff
@@ -1 +1,5 @@
+import { ComposeMailBody } from './types';
 export * from './component';
+export {
+   ComposeMailBody,
```
Comment (bhavyasf) [inline services/in-mail-service/src/index.ts:4]: Will be using in vmsl for type

File: services/in-mail-service/src/controllers/collector.controller.ts (hunk 4 above)
Comment (bhavyasf) [reply services/in-mail-service/src/controllers/collector.controller.ts:63]: item used as getModelSchemaRef.

File: services/in-mail-service/src/controllers/collector.controller.ts (hunk 5 above)
Comment (bhavyasf) [reply services/in-mail-service/src/controllers/collector.controller.ts:250]: others used as filter.

**Comment by samarpan-b** [comment]
rebase please @bhavyasf

File: services/in-mail-service/src/controllers/collector.controller.ts [hunk 15]
```diff
@@ -0,0 +1,357 @@
...
+      threadId,
+      party:
+        process.env.INMAIL_IDENTIFIER_TYPE === 'user'
+          ? this.user.id
+          : this.user.email,
+    });
+    if (!count) {
+      throw new HttpErrors.BadRequest('Group not found');
+    }
+    const threadFilter: Filter<Thread> = {
```
Comment (samarpan-b) [inline services/in-mail-service/src/controllers/collector.controller.ts:92]: is it working ?

File: services/in-mail-service/src/controllers/collector.controller.ts [hunk 16]
```diff
@@ -0,0 +1,357 @@
...
+      };
+    } else {
+      Object.assign(filterThread.where, {
+        id: {inq: threadIds},
+      });
+    }
+
+    const threads = await this.threadRepository.find(filterThread);
+    return {
+      version,
```
Comment (samarpan-b) [inline services/in-mail-service/src/controllers/collector.controller.ts:286]: whats the purpose of version ? It seems to be unused.

File: services/in-mail-service/src/models/id-array.model.ts [hunk 17]
```diff
@@ -0,0 +1,20 @@
+// array containing array of Ids.
+import {property, Entity} from '@loopback/repository';
+
+export class IdArrays extends Entity {
```
Comment (samarpan-b) [inline services/in-mail-service/src/models/id-array.model.ts:4]: @model missing

File: services/in-mail-service/src/models/id-response.model.ts [hunk 18]
```diff
@@ -0,0 +1,17 @@
+import {Entity, property} from '@loopback/repository';
+
+export class IdResponse extends Entity {
```
Comment (samarpan-b) [inline services/in-mail-service/src/models/id-response.model.ts:3]: @model missing

**Review by samarpan-b — CHANGES_REQUESTED** [review]

File: services/in-mail-service/src/controllers/collector.controller.ts (hunk 16 above)
Comment (bhavyasf) [reply services/in-mail-service/src/controllers/collector.controller.ts:286]: API version as discussed with @hitesh-sourcefuse

File: services/in-mail-service/src/controllers/collector.controller.ts (hunk 15 above)
Comment (bhavyasf) [reply services/in-mail-service/src/controllers/collector.controller.ts:92]: yes. tested locally

File: services/in-mail-service/src/models/id-array.model.ts (hunk 17 above)
Comment (bhavyasf) [reply services/in-mail-service/src/models/id-array.model.ts:4]: ok

File: services/in-mail-service/src/models/id-response.model.ts (hunk 18 above)
Comment (bhavyasf) [reply services/in-mail-service/src/models/id-response.model.ts:3]: ok

**Comment by samarpan-b** [comment]
can u check why CI check is failing ?

File: services/in-mail-service/src/models/id-response.model.ts (hunk 18 above)
Comment (samarpan-b) [reply services/in-mail-service/src/models/id-response.model.ts:6]: is there a table name like this ? if yes, then its incorrect format.

**Review by samarpan-b — CHANGES_REQUESTED** [review]

**Pull request merged**